File: platform/openapi/src/operations/agents.ts
/**
 * Agent operations — list, fetch, register, promote, version pinning,
 * privacy-tier check.
 */

import {
  CheckAgentResponse,
  GetAgentResponse,
  ListAgentVersionsResponse,
  ListAgentsResponse,
  PromoteAgentRequest,
  PromoteAgentResponse,
  PromoteRegisteredAgentRequest,
  PromoteRegisteredAgentResponse,
  RegisterAgentJsonRequest,
  RegisterAgentResponse,
} from '@aldo-ai/api-contract';
import type { OpenAPIRegistry } from '../registry.js';
import {
  Resp401,
  Resp403,
  Resp404,
  Resp422,
  Resp422PrivacyTier,
  SECURITY_BOTH,
  jsonResponse,
  pathParam,
} from './_shared.js';

export function registerAgentOperations(reg: OpenAPIRegistry): void {
  reg.registerTag(
    'Agents',
    'Agent specs (data, versioned, eval-gated). Per CLAUDE.md non-negotiable #4.',
  );

  reg.registerPath({
    method: 'get',
    path: '/v1/agents',
    summary: 'List all agents in the current tenant',
    description:
      'Lists every agent visible to the calling tenant with its current version and metadata.',
    tags: ['Agents'],
    security: SECURITY_BOTH,
    responses: { '200': jsonResponse('Agents list.', ListAgentsResponse), '401': Resp401() },
  });

  reg.registerPath({
    method: 'post',
    path: '/v1/agents',
    summary: 'Register a new agent version',
    description:
      'Creates a new agent version from a YAML body. Re-registering the same name appends a new version row; the current version pointer is unchanged unless `setCurrent: true`.',
    tags: ['Agents'],
    security: SECURITY_BOTH,
    request: {
      required: true,
      content: { 'application/json': { schema: RegisterAgentJsonRequest } },
    },
    responses: {
      '200': jsonResponse('Agent registered.', RegisterAgentResponse),
      '401': Resp401(),
      '403': Resp403(),
      '422': Resp422(),
    },
  });

  reg.registerPath({
    method: 'get',
    path: '/v1/agents/{name}',
    summary: 'Fetch a single agent by name',
    description: "Returns the agent's currently-promoted version with its full spec.",
    tags: ['Agents'],
    security: SECURITY_BOTH,
    parameters: [pathParam('name', 'Agent name (slug).', '<agent-name>')],
    responses: {
      '200': jsonResponse('Agent detail.', GetAgentResponse),
      '401': Resp401(),
      '404': Resp404('Agent'),
    },
  });

  reg.registerPath({
    method: 'delete',
    path: '/v1/agents/{name}',
    summary: 'Delete an agent and all its versions',
    description:
      'Hard-deletes the agent record. Past runs that reference this agent retain their snapshot.',
    tags: ['Agents'],
    security: SECURITY_BOTH,
    parameters: [pathParam('name', 'Agent name.', '<agent-name>')],
    responses: { '204': { description: 'Deleted.' }, '401': Resp401(), '404': Resp404('Agent') },
  });

  reg.registerPath({
    method: 'get',
    path: '/v1/agents/{name}/versions',
    summary: 'List all versions of an agent',
    description: 'Returns every recorded version of the agent in registration order.',
    tags: ['Agents'],
    security: SECURITY_BOTH,
    parameters: [pathParam('name', 'Agent name.', '<agent-name>')],
    responses: {
      '200': jsonResponse('Version list.', ListAgentVersionsResponse),
      '401': Resp401(),
      '404': Resp404('Agent'),
    },
  });

  reg.registerPath({
    method: 'get',
    path: '/v1/agents/{name}/versions/{version}',
    summary: 'Fetch a specific agent version',
    description: 'Returns the spec exactly as registered for that version.',
    tags: ['Agents'],
    security: SECURITY_BOTH,
    parameters: [
      pathParam('name', 'Agent name.', '<agent-name>'),
      pathParam('version', 'Version string (e.g. `v3`).', '<version>'),
    ],
    responses: {
      '200': jsonResponse('Version detail.', GetAgentResponse),
      '401': Resp401(),
      '404': Resp404('Agent version'),
    },
  });

  reg.registerPath({
    method: 'post',
    path: '/v1/agents/{name}/check',
    summary: 'Run pre-flight privacy-tier + capability check',
    description:
      "Resolves the agent's privacy tier + capability requirements against the live model catalog WITHOUT executing the agent. Surfaces routing audit so an operator can debug `privacy_tier_unroutable` failures before kicking off a run.",
    tags: ['Agents'],
    security: SECURITY_BOTH,
    parameters: [pathParam('name', 'Agent name.', '<agent-name>')],
    responses: {
      '200': jsonResponse('Routing trace.', CheckAgentResponse),
      '401': Resp401(),
      '404': Resp404('Agent'),
      '422': Resp422PrivacyTier(),
    },
  });

  reg.registerPath({
    method: 'post',
    path: '/v1/agents/{name}/promote',
    summary: 'Promote a YAML body as the new current version',
    description:
      'Validates + persists the YAML and atomically points the current version pointer to it (eval-gate enforced server-side).',
    tags: ['Agents'],
    security: SECURITY_BOTH,
    parameters: [pathParam('name', 'Agent name.', '<agent-name>')],
    request: { required: true, content: { 'application/json': { schema: PromoteAgentRequest } } },
    responses: {
      '200': jsonResponse('Promoted.', PromoteAgentResponse),
      '401': Resp401(),
      '403': Resp403(),
      '422': Resp422(),
    },
  });

  reg.registerPath({
    method: 'post',
    path: '/v1/agents/{name}/set-current',
    summary: 'Set the current version pointer to an existing version',
    description:
      'No new content; just pivots the `current_version` to a previously-registered version. Used to roll back an unhealthy promotion.',
    tags: ['Agents'],
    security: SECURITY_BOTH,
    parameters: [pathParam('name', 'Agent name.', '<agent-name>')],
    request: {
      required: true,
      content: { 'application/json': { schema: PromoteRegisteredAgentRequest } },
    },
    responses: {
      '200': jsonResponse('Pointer updated.', PromoteRegisteredAgentResponse),
      '401': Resp401(),
      '403': Resp403(),
      '404': Resp404('Agent version'),
    },
  });
}

File: sdks/python/src/aldo_ai/resources/agents.py
"""``/v1/agents`` — register, list, get, promote."""

from __future__ import annotations

from typing import Any

from ..types import (
    AgentDetail,
    ListAgentsResponse,
    PromoteRegisteredAgentResponse,
    RegisterAgentResponse,
)
from ._base import _Resource


class AgentsResource(_Resource):
    """Operations on agent specs.

    LLM-agnostic: register accepts the YAML spec verbatim; the server
    parses it through ``@aldo-ai/registry``.
    """

    def list(
        self,
        *,
        team: str | None = None,
        owner: str | None = None,
        cursor: str | None = None,
        limit: int = 50,
    ) -> ListAgentsResponse:
        body = self._sync_t().request(
            "GET",
            "/v1/agents",
            params={"team": team, "owner": owner, "cursor": cursor, "limit": limit},
        )
        return ListAgentsResponse.model_validate(body)

    async def alist(
        self,
        *,
        team: str | None = None,
        owner: str | None = None,
        cursor: str | None = None,
        limit: int = 50,
    ) -> ListAgentsResponse:
        body = await self._async_t().request(
            "GET",
            "/v1/agents",
            params={"team": team, "owner": owner, "cursor": cursor, "limit": limit},
        )
        return ListAgentsResponse.model_validate(body)

    def get(self, name: str) -> AgentDetail:
        body = self._sync_t().request("GET", f"/v1/agents/{name}")
        return AgentDetail.model_validate(body["agent"])

    async def aget(self, name: str) -> AgentDetail:
        body = await self._async_t().request("GET", f"/v1/agents/{name}")
        return AgentDetail.model_validate(body["agent"])

    def register(self, *, spec_yaml: str) -> RegisterAgentResponse:
        body = self._sync_t().request(
            "POST",
            "/v1/agents",
            json_body={"specYaml": spec_yaml},
        )
        return RegisterAgentResponse.model_validate(body)

    async def aregister(self, *, spec_yaml: str) -> RegisterAgentResponse:
        body = await self._async_t().request(
            "POST",
            "/v1/agents",
            json_body={"specYaml": spec_yaml},
        )
        return RegisterAgentResponse.model_validate(body)

    def promote(self, name: str, *, version: str) -> PromoteRegisteredAgentResponse:
        """Eval-gated promote — runs the agent's eval gate suites first.

        See :meth:`set_current` for an unconditional pointer flip.
        """
        body = self._sync_t().request(
            "POST",
            f"/v1/agents/{name}/promote",
            json_body={"version": version},
        )
        return PromoteRegisteredAgentResponse.model_validate(body)

    async def apromote(self, name: str, *, version: str) -> PromoteRegisteredAgentResponse:
        body = await self._async_t().request(
            "POST",
            f"/v1/agents/{name}/promote",
            json_body={"version": version},
        )
        return PromoteRegisteredAgentResponse.model_validate(body)

    def set_current(self, name: str, *, version: str) -> PromoteRegisteredAgentResponse:
        body = self._sync_t().request(
            "POST",
            f"/v1/agents/{name}/set-current",
            json_body={"version": version},
        )
        return PromoteRegisteredAgentResponse.model_validate(body)

    async def aset_current(self, name: str, *, version: str) -> PromoteRegisteredAgentResponse:
        body = await self._async_t().request(
            "POST",
            f"/v1/agents/{name}/set-current",
            json_body={"version": version},
        )
        return PromoteRegisteredAgentResponse.model_validate(body)

    def check(self, name: str) -> Any:
        """Operator dry-run; returns the routing trace as a dict."""
        return self._sync_t().request("POST", f"/v1/agents/{name}/check")

    async def acheck(self, name: str) -> Any:
        return await self._async_t().request("POST", f"/v1/agents/{name}/check")
